backend/src/routes/api/groups-config/groupsConfigUtil.ts

```diff
@@ -0,0 +1,192 @@
+    .filter((group) => group.enabled)
+    .map((group) => group.name);
+  const userConfig = groupConfigUpdated.userGroups
+    .filter((group) => group.enabled)
+    .map((group) => group.name);
+
+  if (adminConfig.length === 0 || userConfig.length === 0) {
+    const error = createError(403, 'Error, groups cannot be empty');
+    throw error;
+  }
+  try {
+    const configMapDataUpdated: GroupConfigMapData = {
+      admin_groups: adminConfig.toString(),
+      allowed_groups: userConfig.toString(),
+    };
+    const groupsConfigName = await getGroupsConfigMapName(coreV1Api, namespace);
+    const groupsData = await updateGroupsConfigMapData(
+      coreV1Api,
+      namespace,
+      groupsConfigName,
+      configMapDataUpdated,
+    );
+    const groupsProcessed = processGroupData(groupsData);
+    const groupsCluster = await getAllGroups(customObjectsApi);
+    const updatedConfig = processGroupConfig(groupsProcessed, groupsCluster);
+    if (
+      groupsData.admin_groups !== updatedConfig.groupsConfigMapData.admin_groups ||
+      groupsData.allowed_groups !== updatedConfig.groupsConfigMapData.allowed_groups
+    ) {
+      await updateGroupsConfigMapData(
+        coreV1Api,
+        namespace,
+        groupsConfigName,
+        updatedConfig.groupsConfigMapData,
+      );
+    }
+    return {
+      success: updatedConfig.groupsConfig,
+      error: null,
+    };
+  } catch (e) {
+    fastify.log.error('Error updating group configuration' + e.toString());
+    const error = createError(500, 'Error updating group configuration');
+    throw error;
+  }
+};
+
+const processGroupData = (groupsData: GroupConfigMapData): GroupConfigMapDataProcessed => {
+  const adminGroupsList = groupsData.admin_groups.split(',');
+  const userGroupList = groupsData.allowed_groups.split(',');
+
+  return {
+    adminGroup: adminGroupsList,
+    allowedGroup: userGroupList,
+  };
+};
+
+const processGroupConfig = (
+  groupsDataProcessed: GroupConfigMapDataProcessed,
+  groups: string[],
+): { groupsConfig: GroupsConfig; groupsConfigMapData: GroupConfigMapData } => {
+  const adminGroupsConfig = groups.map((group, index) => ({
+    id: index,
+    name: group,
+    enabled: groupsDataProcessed.adminGroup.includes(group),
+  }));
+  const userGroupsConfig = groups.map((group, index) => ({
+    id: index,
+    name: group,
+    enabled: groupsDataProcessed.allowedGroup.includes(group),
+  }));
+
+  userGroupsConfig.push({
+    id: userGroupsConfig.length,
+    name: SYSTEM_AUTHENTICATED,
+    enabled: groupsDataProcessed.allowedGroup.includes(SYSTEM_AUTHENTICATED),
+  });
+
+  const groupsConfig: GroupsConfig = {
+    adminGroups: adminGroupsConfig,
+    userGroups: userGroupsConfig,
+  };
+
+  const missingAdminGroups = groupsDataProcessed.adminGroup.filter(
+    (group) => !groups.includes(group),
+  );
+  const missingUserGroups = groupsDataProcessed.allowedGroup.filter(
+    (group) => !groups.includes(group) && group !== SYSTEM_AUTHENTICATED,
+  );
```

Comment:
These filters and the if-statements below actually feel very tied together and causing some DRY issues.

Try just making a predicate function for the filter and returning the error:
```
  const groupsConfig: GroupsConfig = {
    adminGroups: adminGroupsConfig,
    userGroups: userGroupsConfig,
    errorAdmin: getError(groupsDataProcessed.adminGroup, (group) => !groups.includes(group)),
    errorUser: getError(
      groupsDataProcessed.allowedGroup,
      (group) => !groups.includes(group) && group != SYSTEM_AUTHENTICATED,
    ),
  };
```

Where you have a function like:
```
  const getError = (array: string[], predicate: (group: string) => boolean): string | undefined => {
    const missingItems = array.filter(predicate);
    if (missingItems.length === 0) return undefined;

    return `The group${missingItems.length === 1 ? '' : 's'} ${missingItems.join(
      ', ',
    )} no longer exists in OpenShift and has been removed from the selected group list.`;
  };
```

Reply:
Yeah, you're absolutely right, it happened when I was adding features on top of features for edge-cases, I should've refactored the code before opening the PR 😅